helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java

```diff
@@ -1285,4 +1264,87 @@ public void handleSessionEstablishmentError(Throwable error) throws Exception {
+            .setZkRealmShardingKey(shardingKey)
+            .setSessionTimeout(_sessionTimeout).build();
+
+    RealmAwareZkClient.RealmAwareZkClientConfig clientConfig =
+        new RealmAwareZkClient.RealmAwareZkClientConfig();
+
+    clientConfig.setZkSerializer(zkSerializer)
+        .setConnectInitTimeout(_connectionInitTimeout)
+        .setMonitorType(_instanceType.name())
+        .setMonitorKey(_clusterName)
+        .setMonitorInstanceName(_instanceName)
+        .setMonitorRootPathOnly(isMonitorRootPathOnly());
+
+    RealmAwareZkClient newClient;
+    if (_instanceType == InstanceType.ADMINISTRATOR) {
+      newClient = buildSharedZkClient(connectionConfig, clientConfig);
+    } else {
+      newClient = buildDedicatedZkClient(connectionConfig, clientConfig);
+    }
+
+    return newClient;
+  }
+
+  private RealmAwareZkClient buildSharedZkClient(
+      RealmAwareZkClient.RealmAwareZkConnectionConfig connectionConfig,
+      RealmAwareZkClient.RealmAwareZkClientConfig clientConfig) {
+    RealmAwareZkClient zkClient;
+    // TODO: change exception type
+    try {
+      zkClient =
+          SharedZkClientFactory.getInstance().buildZkClient(connectionConfig, clientConfig);
+    } catch (IllegalArgumentException e) {
+      LOG.warn("Not able to build realm-aware shared ZK client for sharding key: {}, caused by: {}"
+              + " Fallback to HelixZkClient.", connectionConfig.getZkRealmShardingKey(),
+          e.getMessage());
+
+      // Fallback to HelixZkClient
+      HelixZkClient.ZkConnectionConfig helixZkConnectionConfig =
+          new HelixZkClient.ZkConnectionConfig(_zkAddress);
+      helixZkConnectionConfig.setSessionTimeout(connectionConfig.getSessionTimeout());
+      zkClient = SharedZkClientFactory.getInstance()
+          .buildZkClient(helixZkConnectionConfig, clientConfig.createHelixZkClientConfig());
+    }
+
+    return zkClient;
+  }
+
+  private RealmAwareZkClient buildDedicatedZkClient(
+      RealmAwareZkClient.RealmAwareZkConnectionConfig connectionConfig,
+      RealmAwareZkClient.RealmAwareZkClientConfig clientConfig) {
+    RealmAwareZkClient zkClient;
+    // TODO: change exception type
+    try {
+      zkClient =
+          DedicatedZkClientFactory.getInstance().buildZkClient(connectionConfig, clientConfig);
+    } catch (IllegalArgumentException e) {
+      LOG.warn("Not able to build realm-aware Dedicated ZK client for sharding key: {}, caused by: {}"
+              + " Fallback to HelixZkClient.", connectionConfig.getZkRealmShardingKey(),
+          e.getMessage());
+
+      // Fallback to HelixZkClient
+      HelixZkClient.ZkConnectionConfig helixZkConnectionConfig =
+          new HelixZkClient.ZkConnectionConfig(_zkAddress);
+      helixZkConnectionConfig.setSessionTimeout(connectionConfig.getSessionTimeout());
+      zkClient = DedicatedZkClientFactory.getInstance()
+          .buildZkClient(helixZkConnectionConfig, clientConfig.createHelixZkClientConfig());
+    }
```

Comment:
Do you think we could refactor these two methods to reduce duplicate logic? You could take in the factory in the parameter and switch back and forth between dedicated and shared that way.

Reply:
I tried. We have 2 options:
1. As you said, we can pass in the factory as a parameter. But we have to make `HelixZkClientFactory` public. Shall we do this?
`public abstract class HelixZkClientFactory implements RealmAwareZkClientFactory`

2. It is the implemented as I do. Keep`HelixZkClientFactory` as it is. I prefer this considering that we don't want HelixZkClientFactory public.